zip-0315.rst

```diff
@@ -1,9 +1,620 @@
+* to manage the privacy implications of these actions.
+
+The goal of this ZIP is to document security and privacy best practices when handling
+funds and transactions as the user's agent. These best practices are intended to
+provide as much privacy as is feasible by default, while still enabling the user's
+desired transactions to occur, and exposing any privacy implications to the user so
+that they have enough information to assess the consequences of their economic actions.
+
+This ZIP covers best practices for:
+
+* what information to display to the user about transactions and balances;
+* how to handle interactions between the ZIP 32 key tree and Unified Addresses;
+* when to use external or internal keys/addresses;
+* sharing addresses and viewing keys;
+* sending and receiving funds (including fee selection);
+* migrating funds between pools.
+
+
+Requirements
+============
+
+User consent
+------------
+
+A wallet typically reveals some information in the process of creating a transaction.
+Which information is revealed depends on the configured wallet privacy policy.
+The guiding principle of these requirements is that users must explicitly consent
+to each instance of information being revealed by the wallet in a transaction.
+
+A user may give blanket consent to reveal a particular kind of information, and
+must also be able to change the configured wallet privacy policy to avoid the
+wallet creating new information leaks of a given type.
+
+The specifications below describe some situations in which blanket consent may be
+inappropriate.
+
+Some varieties of consent may not be revocable, for example if a user chooses to
+share some of their keys.
+
+
+Prompt accessibility of funds
+-----------------------------
+
+Wallets need to take account of two concerns:
+
+* enabling funds to be spent as quickly as possible to reduce latency;
+* waiting for long enough before spending TXOs to ensure that the
+  confirmed-spendable balance is not overestimated, and so can be trusted
+  by the user.
+
+To enable this we define two kinds of TXOs:
+
+* An untrusted TXO is a TXO received from a party that may try to double-spend.
+* A trusted TXO is a TXO received from a party where the wallet trusts a
+  double-spend not to occur, e.g. TXOs created by the wallet's internal TXO
+  handling.
+
+Wallets can then require that untrusted TXOs need more confirmations before
+they become spendable than trusted TXOs. This provides an improved trade-off
+between latency on the one hand, and reliability and safety on the other.
+
+
+Specification
+=============
+
+Long-term storage of funds
+--------------------------
+
+It is RECOMMENDED that wallets only hold funds as shielded in the long-term;
+that is, if a wallet supports receiving transparent funds (or supports
+importing a seed from another wallet that may have done so), then it SHOULD
+either auto-shield or opportunistically shield such funds by default.
```

Comment:
I started this comment yesterday, then canceled it because of the “User consent” section, however, I feel like that might not be sufficient in this case.

I _think_ both auto-shielding and opportunistically shielding tx require `AllowRevealedAmounts`, but in the case of opportunistic shielding, the recipient(s) may have information about the sender (either OOB or via the memo), so revealing the amount that’s being shielded has different (worse) privacy implications than in auto-shielding. So I think it’s important for wallets to make the distinction between user consent for auto-shielding and user consent for opportunistic shielding, rather than being like “user has consented to AllowRevealedAmounts for all shielding.”

Reply:
The recipient in an opportunistic shielding transaction sees more than the amount being shielded. They see a transparent address that they now understand is owned by the same party that is paying them. This can reveal far more than the amount being shielded, including the entire transaction history connected to that transparent address.
For this reason alone, I don't think opportunistic shielding should _ever_ be done. Auto-shielding only.

Reply:
I agree; a wallet should never opportunistically shield in a way that links different transparent addresses together.

If it is opportunistically shielding all funds the wallet holds in a single t-addr, OTOH, then there's no privacy leak. Since it is already obvious that the t-addr belongs to the sending wallet, in that case there is no additional information revealed to any party *by the opportunistic shielding*, regardless of what they already know.